File: src/2300_successful_pairs_of_spells_and_potions.js
function successfulPairs(spells, potions, success) {
    const pairs = [];
    potions.sort((a, b) => a - b);
    const n = potions.length;
    for (let i = 0; i < spells.length; i++) {
        let l = 0;
        let r = potions.length - 1;
        while (l <= r) {
            const m = Math.floor((l + r) / 2);
            if (spells[i] * potions[m] >= success) r = m - 1;
            else l = m + 1;
        }
        pairs.push(l < potions.length ? n - l : 0);
    }
    return pairs;
}

module.exports = { successfulPairs };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(successfulPairs([5, 1, 3], [1, 2, 3, 4, 5], 7), [4, 0, 3]);
}

File: src/2300_successful_pairs_of_spells_and_potions.rs
pub fn successful_pairs(spells: Vec<i32>, mut potions: Vec<i32>, success: i64) -> Vec<i32> {
    let mut pairs = Vec::new();
    potions.sort();
    let n = potions.len();

    for spell in spells {
        let mut l = 0_usize;
        let mut r = potions.len();

        while l < r {
            let m = (l + r) / 2;
            if i64::from(spell) * i64::from(potions[m]) >= success {
                r = m;
            } else {
                l = m + 1;
            }
        }
        if l < potions.len() {
            pairs.push((n - l) as i32);
        } else {
            pairs.push(0);
        }
    }

    pairs
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            successful_pairs(vec![5, 1, 3], vec![1, 2, 3, 4, 5], 7),
            vec![4, 0, 3]
        );
    }
}
